### PR TITLE
fix yield-from throw

### LINKED ISSUE #692: py2 `yield-from` does not propagate exceptions *into* the inner generators.
I've found a concrete example of the difference in semantics between the `yield-from` macro and [PEP-380](https://www.python.org/dev/peps/pep-0380#formal-semantics). When using `generator.throw` in Python 2, `yield-from` does not rethrow the exception in the inner generator:

``` hy
hy 0.10.1 using CPython(default) 2.7.8 on Linux
=> (defn f []
...   (yield 1)
...   (try
...     (yield 2)
...     (catch [e ValueError] (yield 3))))
=> (defn g []
...   (yield-from (f))
...   (assert true))  ; see issue #691
=> (setv x (g))
=> (next x)
1L
=> (next x)
2L
=> (x.throw ValueError)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 257, in g  ; note that ValueError didn't even reach `f`
ValueError
```

In Python 3, `ValueError` reaches `f` and is handled there as expected:

``` hy
hy 0.10.1 using CPython(default) 3.4.2 on Linux
...
=> (next x)
2
=> (x.throw ValueError)
3
```

### PR DESCRIPTION
**Pull request by gilch**
Closes #692.

### CODE REVIEW

**Comment by zackmdavis** [comment]
(Continued from discussion on #692)

It does look to me like this suffers from #691&mdash;

```
(hy2)zmd@ExpectedReturn:~/Code/hy2/hy2$ hy 
hy 0.11.0 using CPython(default) 2.7.6 on Linux
=> (defn fail [] (if some-test (yield-from some-presumptive-generator)))
Traceback (most recent call last):
  File "/home/zmd/Code/hy2/hy2/hy/importer.py", line 44, in ast_compile
    return compile(ast, filename, mode, flags)
SyntaxError: 'return' with argument inside generator (<input>, line 1)
```
The problem seems to be that we sometimes translate a Hy expression in tail position to a Python assign-a-temporary-value-then-return-it maneuver, and mixing `return` and `yield` in the same function doesn't work in Python 2; the compiler ostensibly [tries to deal with this](https://github.com/hylang/hy/blob/fb9a97a7/hy/compiler.py#L2320-L2323), but as our BDFL commented on #691, it would seem to be buggy.

(This pull request might still be a good idea, though! Faithfully translating the "Formal semantics" section of PEP 380 probably has _fewer_ bugs than the existing `yield-from` backport, even if there's another bug elsewhere in the `contains_yield` stuff.)

File: hy/core/macros.hy [hunk 1]
```diff
@@ -214,20 +214,42 @@
...
-                                     (. ~g!e value)
-                                     nil))
-               (break)))))
-           ~g!return))
+    `(do
+       (setv ~g!i (iter ~expr))
+       (try
+         (setv ~g!y (next ~g!i))
+         (except [~g!e StopIteration]
+                 (setv ~g!r (. ~g!e value)))
```
Comment (pyos) [inline hy/core/macros.hy]: This should be `(setv ~g!r (next (iter (. ~g!e args)) nil))`. StopIteration instances don't have the `value` attribute in Python 2.

File: hy/core/macros.hy [hunk 2]
```diff
@@ -214,20 +214,42 @@
...
+               (except [~g!e BaseException]
+                       (setv ~g!x (sys.exc_info))
+                       (try
+                         (setv ~g!m (. ~g!i throw))
+                         (except [AttributeError] (raise ~g!e))
+                         (else
+                           (try
+                             (setv ~g!y (apply ~g!m ~g!x))
+                             (except [~g!e StopIteration]
+                                     (setv ~g!r (. ~g!e value))
```
Comment (pyos) [inline hy/core/macros.hy]: ditto

File: hy/core/macros.hy [hunk 3]
```diff
@@ -214,20 +214,42 @@
...
+                                     (break))))))
+               (else
+                 (try
+                   (if (is ~g!s nil)
+                     (setv ~g!y (next ~g!i))
+                     (setv ~g!y ((. ~g!i send) ~g!s)))
+                   (except [~g!e StopIteration]
+                           (setv ~g!r (. ~g!e value))
```
Comment (pyos) [inline hy/core/macros.hy]: same

**Comment by pyos** [comment]
>mixing return and yield in the same function doesn't work in Python 2

And in Python 3, `return x` inside a generator is equivalent to `raise StopIteration(x)` (the argument is accessible as the `value` attribute on the resulting exception object), so even if yield detection works, the compiler still does the wrong thing. Without this, coroutines would be much less useful.

**Comment by gilch** [comment]
@zackmdavis, I think see why `yield` isn't detected: `defmacro/g!` expands to a `let` form. So the `yield` isn't in the function we think it is in. If I redefine `defmacro/g!` in terms of `setv`, the `yield` detection should work. I think I'll add that to the PR. We were considering getting rid of `let` anyway #1056, and that means removing all of its uses from the core macros.

@pyos, yeah, the [PEP-380 "Formal Sematics" section](https://www.python.org/dev/peps/pep-0380/#formal-semantics) that I translated this from specifically said "Python 3 syntax is used in this section." I figured something would break in Python 2, but I didn't uncover it in my limited initial tests.

When deviating from the direct translation though, I'd prefer to have a unit test to ensure your proposed changes are actually doing what we want. (Ideally we'd test all the branches.) I'm not sure how thorough our current `yield-from` unit tests are, but (if it's not covered already) can you also come up with some concrete examples illustrating the need for these changes? As you pointed out, the compiler is also doing the wrong thing in some cases.

**Comment by pyos** [comment]
@gilch,

>I think see why `yield` isn't detected: `defmacro/g!` expands to a `let` form. So the `yield` isn't in the function we think it is in.

Even so, why does the function that `let` creates contain a `return` when it also contains a `yield`? There's still a bug somewhere else...

>When deviating from the direct translation though, I'd prefer to have a unit test to ensure your proposed changes are actually doing what we want.

Well, it looks like there's already a test that fails due to absense of `value` in Python 2 (check the Travis logs). No test ensures that coroutines/generators return a value in `StopIteration`, though; such a test should do something similar to this Python code: http://ideone.com/hqD1py

**Comment by gilch** [comment]
>Even so, why does the function that let creates contain a return when it also contains a yield? The exception is pretty unambiguous.

I double-checked the expansion of `defmacro/g!`. I miscounted the quotes. While it does use `let` internally, it doesn't appear in the expansion at all, so `yield` wouldn't be wrapped in another function. `defmacro/g!` is not the reason for this problem.

**Comment by gilch** [comment]
A `hy --spy` with the new version. Note the erroneous `return` at the end.
```Python
=> (defn f[x] (raise (StopIteration x))(yield))
def f(x):
    raise StopIteration(x)
    yield
=> (defn g[x] (print (yield-from (f x))))
def g(x):
    :i_1235 = iter(f(x))
    try:
        :y_1236 = next(:i_1235)
        _hy_anon_var_1 = :y_1236
    except StopIteration as :e_1237:
        :r_1238 = next(iter(:e_1237.args), None)
        _hy_anon_var_1 = :r_1238
    else:
        while True:
            try:
                :s_1239 = yield :y_1236
                _hy_anon_var_2 = :s_1239
            except GeneratorExit as :e_1237:
                try:
                    :m_1240 = :i_1235.close
                    _hy_anon_var_3 = :m_1240
                except AttributeError:
                    _hy_anon_var_3 = pass
                else:
                    :m_1240()
                raise :e_1237
                _hy_anon_var_2 = None
            except BaseException as :e_1237:
                :x_1241 = sys.exc_info()
                try:
                    :m_1240 = :i_1235.throw
                    _hy_anon_var_4 = :m_1240
                except AttributeError:
                    raise :e_1237
                    _hy_anon_var_4 = None
                else:
                    try:
                        :y_1236 = :m_1240(*:x_1241)
                        _hy_anon_var_5 = :y_1236
                    except StopIteration as :e_1237:
                        :r_1238 = next(iter(:e_1237.args), None)
                        break
                        _hy_anon_var_5 = None
                _hy_anon_var_2 = _hy_anon_var_4
            else:
                try:
                    if (:s_1239 is None):
                        :y_1236 = next(:i_1235)
                        _hy_anon_var_6 = :y_1236
                    else:
                        :y_1236 = :i_1235.send(:s_1239)
                        _hy_anon_var_6 = :y_1236
                    _hy_anon_var_7 = _hy_anon_var_6
                except StopIteration as :e_1237:
                    :r_1238 = next(iter(:e_1237.args), None)
                    break
                    _hy_anon_var_7 = None
    return print(:r_1238)
Traceback (most recent call last):
  File "c:\users\me\code\repos\hy\hy\importer.py", line 44, in ast_compile
    return compile(ast, filename, mode, flags)
SyntaxError: 'return' with argument inside generator (<input>, line 1)
=>
```

**Comment by Kodiologist** [comment]
@gilch How are we doing here?

**Comment by gilch** [comment]
There are conflicts now, but this change is pretty localized, so it shouldn't be hard to resolve. 

This is waiting on #739, or something. Since there's currently no way to suppress that spurious `return` at the end of the Python expansion, this will always fail. We need a way to suppress that implicit return. We haven't done so (so far) because in most cases, you can simply return `None` and it's effectively the same as if you had no return statement, but that doesn't work here, because it's a generator. `return` must not appear in the output at all. I suggested adjusting the compiler to use a `pass` in the tail position for this purpose, but that's not the only approach we've discussed.

**Comment by Kodiologist** [comment]
Since there's been no progress on this PR for about a year, I'm going to close it. Feel free to reopen it once work has resumed.

**Comment by gilch** [comment]
We might drop Python 2 support before that happens. Then this will be obsolete. Like I said before, it's waiting on another problem, (the spurious `return`) which might still be trouble after that. I'll leave this closed for now.

**Pull request closed without merging**